[llvm][MC][ARM] Don't autoresolve fixups

eleanor-arm opened the pull request:
Removes logic that caused some fixups to be marked as resolved in the
assembler without actually resolving them. Assembler must either resolve
the fixup, reject the code (error out) or defer the resolution to the
linker. In general assembler can resolve offsets in pcrel instructions
if the symbol referred to is in the same section and it cannot make
assumptions about relative position of sections. For example, when LDRD
instruction in arm mode is encountered, fixup_arm_pcrel_10_unscaled is
raised. Prior to https://github.com/llvm/llvm-project/pull/72873/ the
assembler would always mark it as resolved without generating a
relocation. The resulting code would likely be incorrect whenever the
label referred to is in a different section.

This patch finishes the series to prevent incorrect code being generated
for pcrel instructions referring to labels in different sections. This
kind of assembly code is very rare and most likely a user error, so both
options (relocation or error) are acceptable. In previous patches this
was resolved by adding relocations. Here, for VLDR instructions an error
is generated as relocation does not exist for Thumb mode and we wanted
the tool's behaviour to be consistent across modes. In the LDRD case,
Thumb mode does not have a relocation and errors out, but LDRD in Arm
mode generates R_ARM_LDRS_PC_G0 relocation because its fixup kind is
shared with other instructions.

It also fixed the case when ADR is used in the big-endian mode, which is
not covered by the ADR patch.

Patch series:
https://github.com/llvm/llvm-project/pull/72873 - LDRx
https://github.com/llvm/llvm-project/pull/73834 - ADR 
this PR - LDRD and VLDR

eleanor-arm committed:
[llvm][MC][ARM] Don't autoresolve fixups

Removes logic that caused some fixups to be marked as resolved in the
assembler without actually resolving them. Sometimes to resolve these
fixups a relocation was needed, which was not generated, resulting in
invalid code. Now in situations when the fixup cannot be resolved by the
assembler, either a relocation is generated, or an error is produced.

This was partially addressed previously
(https://github.com/llvm/llvm-project/pull/72873/,
https://github.com/llvm/llvm-project/pull/73834) specifically for LDRx
and ADR instructions, This patch expands it LDRD and VLDR, which should
cover all instructions affected.

MaskRay commented on llvm/test/MC/ARM/pcrel-global.s:
```diff
@@ -7,18 +7,14 @@
 
-.globl foo
-foo:
-vldr d0, foo     @ arm_pcrel_10
```
Why is this removed?

MaskRay commented on llvm/test/MC/ARM/pcrel-thumb-ldrd-relocs.s:
```diff
@@ -0,0 +1,6 @@
+@ RUN: not llvm-mc -filetype=obj -o /dev/null %s 2>&1 -triple=thumbv7   | FileCheck %s
+@ RUN: not llvm-mc -filetype=obj -o /dev/null %s 2>&1 -triple=thumbebv7 | FileCheck %s
+
+   ldrd r0, r1, foo
```
Is this related to positive tests in another test file? If true, we can merge the test files by utilizing `.ifdef ERR` and `--defsym ERR=1`. Grep will a few examples.

eleanor-arm replied:
Thanks for the tip. I've merged `pcrel-thumb-ldrd-relocs.s` and `pcrel-arm-ldrd-relocs.s` into `pcrel-ldrd-diff-section.s`

MaskRay commented on llvm/test/MC/ARM/pcrel-thumb-ldrd-relocs.s:
```diff
@@ -0,0 +1,6 @@
+
+   ldrd r0, r1, foo
+
+@ CHECK: :[[#@LINE-2]]:4: error: unsupported relocation type
```
The CHECK is usually immediately before or after the instruction, so that we can use -1 +1 instead of -N +N

eleanor-arm replied:
As this check has been merged into `pcrel-ldrd-diff-section.s` I found it quite hard to read when it was placed next to the instructions, with all other checks in the same file. I placed it at the of the file next to the other checks.

MaskRay commented on llvm/test/MC/ARM/pcrel-vldr-diff-section.s:
```diff
@@ -0,0 +1,15 @@
+@ CHECK: :[[#@LINE-4]]:1: error: unsupported relocation type
+@ CHECK: :[[#@LINE-4]]:1: error: unsupported relocation type
+@ CHECK: :[[#@LINE-4]]:1: error: unsupported relocation type
+
```
no blank line at the end

eleanor-arm replied:
Update: As I was making a change anyway, I also removed the blank lines.

MaskRay replied:
Leaving a comment at the end of file is very uncommon in the repository. A small set of very old tests may do this.

MaskRay commented:
> Removes logic that caused some fixups to be marked as resolved in the assembler without actually resolving them. 

Maybe worth updating the description to include an example like
```
xxx  // previously incorrectly resolved, now ...
```

eleanor-arm committed:
fixup! [llvm][MC][ARM] Don't autoresolve fixups

eleanor-arm reviewed (commented):
Review comments addressed.

eleanor-arm committed:
amend! [llvm][MC][ARM] Don't autoresolve fixups

[llvm][MC][ARM] Don't autoresolve fixups

Removes logic that caused some fixups to be marked as resolved in the
assembler without actually resolving them. Assembler must either resolve
the fixup, reject the code (error out) or defer the resolution to the
linker. In general assembler can resolve offsets in pcrel instructions
if the symbol referred to is in the same section and it cannot make
assumptions about relative position of sections.  For example, when LDRD
instruction in arm mode is encountered,  fixup_arm_pcrel_10_unscaled is
raised.  Prior to https://github.com/llvm/llvm-project/pull/72873/ the
assembler would always mark it as resolved without generating a
relocation. The resulting code would likely be incorrect whenever the
label referred to is in a different section.

This patch finishes the series to prevent incorrect code being generated
for pcrel instructions referring to labels in different sections. This
kind of assembly code is very rare and most likely a user error, so both
options (relocation or error) are acceptable.  Previously this was
resolved by adding relocations. Here, for VLDR instructions an error is
generated as relocation does not exist for Thumb mode and  we wanted the
tool's behaviour to be consistent across modes. In the LDRD case, Thumb
mode  does not have a relocation and errors out, but LDRD in Arm mode
generates relocation because its fixup kind is shared with other
instructions.

Patch series:
https://github.com/llvm/llvm-project/pull/72873 - LDRx
https://github.com/llvm/llvm-project/pull/73834 - ADR
this PR -  LDRD and VLDR

eleanor-arm commented:
I also elaborated more in the description.

MaskRay commented on llvm/test/MC/ARM/pcrel-ldrd-diff-section.s:
```diff
@@ -0,0 +1,37 @@
+@ RUN: not llvm-mc -filetype=obj --defsym=ERR=1 -o /dev/null %s 2>&1 -triple=thumbebv7 | FileCheck %s --check-prefix=ERR
+@ RUN: llvm-mc -filetype=obj -triple=armv7 %s -o %t
+@ RUN: llvm-readelf -r %t | FileCheck %s --check-prefix=ARM
+@ RUN: llvm-objdump -d --triple=armv7 %t | FileCheck %s --check-prefix=ARM_ADDEND
```
`llvm-objdump -d -r --no-show-raw-insn` dumps inline relocation when printing disassembly code. This allows to remove `llvm-readelf`

eleanor-arm replied:
I know this is another way to do it but is it really better? I think using Filecheck can easily get confusing, so matching two simple strings seems like a better idea, in general, compared to matching one more complicated one (especially since the tools' output is in flux and we don't want big parts of tests suddenly failing because of a change in the output format of a tool). I copied that idea from other Arm assembly tests in there.

MaskRay commented on llvm/test/MC/ARM/pcrel-ldrd-same-section.s:
```diff
@@ -0,0 +1,39 @@
+@ RUN: llvm-readelf -r %t | FileCheck %s --check-prefix=RELOC
+@ RUN: llvm-objdump -d --triple=armv7 %t | FileCheck %s --check-prefix=ARM_OFFSET
+
+@ RUN: llvm-mc -filetype=obj -o %t %s -triple=armebv7
```
Does it need to be replicated for {arm,thunb}x{le,be}?

Testing (arm,le) + (thumb,be) may suffice, if you think thumb is worth separate testing (in some cases it's not).

eleanor-arm replied:
I don't think it hurts. When/if these start failing it might be helpful with debug.  Internally all these 4 go through different logic to resolve the fixup. Arm/Thumb have separate fixup kinds and and BE/LE have different way of assigning FixupKindInfo in `ARMAsmBackend.cpp`.

MaskRay reviewed (commented):
Please can you edit the description to not indent the whole message by 4? GitHub renders the whole message as code block. Can you also write the exact relocation type names in the description so that it's easier to search? The fixup kinds are internal and users don't grep them for a behavior change.

MaskRay commented on llvm/test/MC/ARM/pcrel-vldr-diff-section.s:
```diff
@@ -0,0 +1,14 @@
+@ RUN: not llvm-mc -filetype=obj -o /dev/null %s 2>&1 -triple=armv8.2a-eabi     | FileCheck %s
```
same-section and diff-section tests can be in one file. `llvm/test/MC/ARM/pcrel-vldr-same-section.s` contains a lot of details that seem unrelated to the main purpose of the patch and should be reduced. There are also lots of instructions that may contain duplicates.

eleanor-arm replied:
How would you put them in the same test? Vldr_same/diff_section.s  have slightly different code, one compiles, the other one doesn't. If I merge them it will fail as a whole.

eleanor-arm replied:
As for the "unrelated details", I think you're referring to the checks of the addend values? And/or the symbol+imm checks?  We've done something similar in all the tests related to pcrel relocations. It was recommended to me as a good extra check by @smithp35 .  Here I just added a few more because we had a discussion on another review about whether an issue could be related to a pcrel instruction referring to a symbol before it rather than after. It wasn't but still I covered both cases from then on as to make it clearer.

eleanor-arm committed:
Fix "ldr" in the disassembly for vldr

MaskRay commented on llvm/test/MC/ARM/pcrel-ldrd-diff-section.s:
```diff
@@ -0,0 +1,37 @@
+    .word 0x99aabbcc, 0xddeeff00
+
+.ifdef ERR
+  @ ERR:[[#@LINE-14]]:5: error: unsupported relocation type
```
`-14` is too far way from the instruction. Move this closer to the instructions so that it's easier to fix an issue if a new line is added.

ellie118 replied:
Done

eleanor-arm replied:
Done

eleanor-arm committed:
Move ERR check in pcrel-ldrd-diff-section.s closer to instruction
